src/host/_stream.cpp

```diff
@@ -1034,199 +1033,148 @@ constexpr unsigned int LOCAL_BUFFER_SIZE = 100;
 {
     try
     {
-        const CONSOLE_INFORMATION& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
         // Ensure output variables are initialized.
         read = 0;
         waiter.reset();
 
-        bool fLeadByteCaptured = false;
-        bool fLeadByteConsumed = false;
-
-        LockConsole();
-        auto Unlock = wil::scope_exit([&] { UnlockConsole(); });
-
-        if (buffer.size() == 0)
+        if (buffer.empty())
         {
             return S_OK;
         }
 
-        const auto codepage = gci.OutputCP;
-
-        // Convert our input parameters to Unicode
-        std::unique_ptr<wchar_t[]> wideCharBuffer{ nullptr };
-        static Utf8ToWideCharParser parser{ gci.OutputCP };
+        LockConsole();
+        auto unlock{ wil::scope_exit([&] { UnlockConsole(); }) };
 
-        // update current codepage in case it was changed from last time
-        // this was called. We do this outside the UTF-8 check because the parser drops its state
-        // when the codepage changes.
-        parser.SetCodePage(gci.OutputCP);
+        auto& screenInfo{ context.GetActiveBuffer() };
+        const auto& consoleInfo{ ServiceLocator::LocateGlobals().getConsoleInformation() };
+        const auto codepage{ consoleInfo.OutputCP };
+        auto leadByteCaptured{ false };
+        auto leadByteConsumed{ false };
+        std::wstring wstr{};
+        static til::u8state u8State{};
 
-        SCREEN_INFORMATION& ScreenInfo = context.GetActiveBuffer();
-        wchar_t* pwchBuffer;
-        size_t cchBuffer;
+        // Convert our input parameters to Unicode
         if (codepage == CP_UTF8)
         {
-            wideCharBuffer.release();
-            unsigned int charCount;
-            unsigned int charsConsumed;
-            unsigned int charsGenerated;
-            RETURN_IF_FAILED(SizeTToUInt(buffer.size(), &charCount));
-            RETURN_IF_FAILED(parser.Parse(reinterpret_cast<const byte*>(buffer.data()),
-                                          charCount,
-                                          charsConsumed,
-                                          wideCharBuffer,
-                                          charsGenerated));
-
-            pwchBuffer = reinterpret_cast<wchar_t*>(wideCharBuffer.get());
-            cchBuffer = charsGenerated;
-            read = charsConsumed;
+            RETURN_IF_FAILED(til::u8u16(buffer, wstr, u8State));
+            read = buffer.size();

```

Comment:
@DHowett-MSFT I was struggling with the same concern like you. We may decrease the effective number of bytes used by putting some in the cache and we also may increase the number by prepending the cached to the string. But for the records, the `utf8Parser::Parse()` method behaves the same. If it succeeds, it populates the incoming length as the length consumed. Since I'm not aware of any issues with this (and I didn't want to cause issues) I left this behavior unchanged.